web/models/package.ex

```diff
@@ -115,26 +116,77 @@ defmodule HexWeb.Package do
   end
 
   defp search(query, search) when is_binary(search) do
-    filter =
-      if String.length(search) >= 3 do
-        {:contains, search}
-      else
-        {:equals, search}
-      end
-
-    search(query, filter)
+    if Enum.any?(@search_keys, & String.starts_with?(search, &1)) do
+      parse_search(search)
+      |> Enum.reduce(query, fn({k, v}, q) ->
+        search(k, q, v)
+      end)
+    else
+      search = Utils.safe_search(search)
+      filter =
+        if String.length(search) >= 3 do
+          :contains
+        else
+          :equals
+        end
+      name_search =
+        search
+        |> escape_search
+        |> like_search(filter)
+
+      desc_search = String.replace(search, ~r"\s+"u, " | ")
+
+      from p in query,
+      where: ilike(fragment("?::text", p.name), ^name_search) or
+             fragment("to_tsvector('english', regexp_replace((?->'description')::text, '/', ' ')) @@ to_tsquery('english', ?)", p.meta, ^desc_search)
+    end
   end
 
-  defp search(query, {mode, search}) do
-    name_search = search |> escape_search() |> like_search(mode)
+  defp search(:name, query, search) do
+    from(p in query,
+      where: ilike(fragment("?::text", p.name), ^search))
+  end
 
+  defp search(:description, query, search) do
     desc_search = String.replace(search, ~r"\s+"u, " | ")
+    from(p in query,
+      where: fragment("to_tsvector('english', regexp_replace((?->'description')::text, '/', ' ')) @@ to_tsquery('english', ?)", p.meta, ^desc_search))
+  end
 
-    # without fragment("?::text", var.name) the gin_trgm_ops index will not be used
-    from var in query,
-    where: ilike(fragment("?::text", var.name), ^name_search) or
-           fragment("to_tsvector('english', regexp_replace((?->'description')::text, '/', ' ')) @@ to_tsquery('english', ?)",
-                    var.meta, ^desc_search)
+  defp search(:extra, query, search) do
+    [v | p] =
+    search
+    |> String.split(":")
+    |> Enum.reverse
+    [h | t] = p
+
+    value =
+      extra_value(v)
+
+    extra =
+      extra(t, Map.put(%{}, h, value))
+    from(p in query,
+      where: fragment("?->'extra' @> ?", p.meta, ^extra))
+  end
+
+  defp extra_value(<<"[", value :: binary>>) do
+    value
+    |> String.rstrip(?])
+    |> String.split(",")
+    |> Enum.map(fn(v) ->
+      case Integer.parse(v) do
+        {int, ""} -> int
+        _ -> Utils.safe_search(v)
+      end
+    end)
+  end
+  defp extra_value(v),
+    do: Utils.safe_search(v)
```

Comment:
Do you have a need for it in nerves? Otherwise I would like to keep the implementation simple. If you do have a use case for it we should of course keep it.